// wsapix/dist/server.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wsapix = void 0;
const transport_1 = require("./transport");
const asyncapi_1 = require("./asyncapi");
const types_1 = require("./types");
const channel_1 = require("./channel");
const template_1 = require("./template");
class Wsapix extends channel_1.WsapixChannel {
    /**
     * Wsapix constructor
     * @param transport - wsapix transport
     * @param defaultOptions - channel options
     */
    constructor(transport, defaultOptions) {
        super((defaultOptions === null || defaultOptions === void 0 ? void 0 : defaultOptions.path) || "*", defaultOptions);
        this.transport = transport;
        /**
         * Map of registered routes
         */
        this.channels = new Map();
        this.setTransport(transport);
    }
    /**
     * Create Wsapix instance on Http server
     * @param options - Http server options
     * @param defaultOptions - channel options
     * @returns Wsapix instance
     */
    static WS(options, defaultOptions) {
        const transport = new transport_1.WebsocketTransport(options);
        return new Wsapix(transport, defaultOptions);
    }
    /**
     * Create Wsapix instance on uWebSockets server
     * @param options - uWebsocket server and options
     * @param defaultOptions - channel options
     * @returns Wsapix instance
     */
    static uWS(options, defaultOptions) {
        const transport = new transport_1.uWebsocketTransport(options);
        return new Wsapix(transport, defaultOptions);
    }
    /**
     * Update wsapix transport
     * @param transport - wsapix transport
     */
    setTransport(transport) {
        this.transport = transport;
        this.transport.onConnection(this.onConnected.bind(this));
        this.transport.onMessage(this.onMessage.bind(this));
        this.transport.onDisconnect(this.onDisconnect.bind(this));
        this.transport.onError((error) => this.emit("error", (error === null || error === void 0 ? void 0 : error.message) || ""));
    }
    onConnected(client) {
        const _super = Object.create(null, {
            onConnect: { get: () => super.onConnect }
        });
        return __awaiter(this, void 0, void 0, function* () {
            // check if channel exist
            const channel = this.findChannel(client.path);
            if (!channel) {
                return client.terminate(4000);
            }
            // handle client connection to channel
            _super.onConnect.call(channel, client);
        });
    }
    onMessage(client, data) {
        const _super = Object.create(null, {
            onMessage: { get: () => super.onMessage }
        });
        return __awaiter(this, void 0, void 0, function* () {
            if (!client.channel) {
                return;
            }
            return _super.onMessage.call(client.channel, client, data);
        });
    }
    onDisconnect(client, code, data) {
        if (!client.channel) {
            return;
        }
        super.onDisconnect.call(client.channel, client, code, data);
    }
    /**
     * Generate AsyncApi specification in json format
     * @param params - AsyncApi specification parts (info, servers, defaultContentType, tags, externalDocs)
     * @returns AsyncApi documentation json
     */
    asyncapi(params) {
        const asyncapi = new asyncapi_1.AsyncApiBuilder(params);
        const channels = [this, ...this.channels.values()];
        for (const channel of channels) {
            // parse path params
            const pathParams = {};
            // split pubsub messages
            const pubMessages = [];
            const subMessages = [];
            for (const msg of channel.messages) {
                if (!msg.schema) {
                    continue;
                }
                if (msg.kind === types_1.MessageKind.server) {
                    subMessages.push(msg.schema);
                }
                else {
                    pubMessages.push(msg.schema);
                }
            }
            // add channel
            asyncapi.addChannel(channel.path, pubMessages, subMessages, pathParams);
        }
        return asyncapi.generate();
    }
    /**
     * @public
     * Generate HTML template for AsyncApi documentation
     * @param asyncApiPath - url for AsyncApi json file
     * @param title - Page title
     * @returns - HTML page
     */
    htmlDocTemplate(asyncApiPath, title) {
        return template_1.html(asyncApiPath, title);
    }
    /**
     * Find channel by url
     * @param url
     * @returns channel or undefined
     */
    findChannel(url = "/") {
        const [path] = url.split("?");
        for (const [channelPath, channel] of this.channels) {
            const exp = new RegExp("\/" + channelPath
                .replace(/^\/+|\/+$/g, "")
                .split("/")
                .map((item) => item[0] === "{" && item.slice(-1) === "}" ? "[A-Za-z0-9_.\-]+" : item)
                .join("\/") + "$");
            if (exp.test(path)) {
                return channel;
            }
        }
        return (this.path === url || this.path === "*") ? this : this.channels.get("*");
    }
    /**
     * Register route channel
     * @param path - route path or channel
     *
     * @example
     * Create new route
     * ```ts
     * const chat = wsx.route("/chat", { parser, serializer })
     * ```
     *
     * Register existing channel instance
     * ```ts
     * const chat = new WsapixChannel("/chat", { parser, serializer })
     * wsx.route(chat)
     * ```
     * Route path supports parameters:
     * @example
     * `/rooms/{id}`
     *
     * @param options - channel options
     * @returns Channel instance
     */
    route(path, options) {
        // create channel
        const channel = path instanceof channel_1.WsapixChannel ? path : new channel_1.WsapixChannel(path, options);
        const channelPath = channel.path;
        if (channelPath === this.path || this.channels.has(channelPath)) {
            throw new Error(`Path '${channelPath}' already exist!`);
        }
        this.inherit.call(channel, this);
        this.channels.set(channelPath, channel);
        return channel;
    }
    /**
     * Plug extension
     * @param plugin - Wsapix extention
     * @returns void
     */
    register(plugin) {
        return plugin(this);
    }
    /**
     * Gracefull shutdown
     * @param cb - Callback funxtion
     * @returns promise
     */
    close(cb) {
        return this.transport.close(cb);
    }
}
exports.Wsapix = Wsapix;
