// wsapix/dist/transport/transport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transport = exports.Client = exports.ClientStatus = void 0;
var ClientStatus;
(function (ClientStatus) {
    ClientStatus[ClientStatus["connecting"] = 0] = "connecting";
    ClientStatus[ClientStatus["connected"] = 1] = "connected";
    ClientStatus[ClientStatus["disconnecting"] = 2] = "disconnecting";
    ClientStatus[ClientStatus["disconnected"] = 3] = "disconnected";
})(ClientStatus = exports.ClientStatus || (exports.ClientStatus = {}));
const noop = () => { };
/**
 * Abstract client class
 */
class Client {
    constructor() {
        /**
         * connection headers
         */
        this.headers = {};
        /**
         * client connection status
         */
        this.status = ClientStatus.connecting;
    }
    /**
     * Send message to client
     * @param data - payload
     * @param cb - callback on error/complete
     * @returns promise
     */
    send(data, cb) {
        return this._send(data, cb);
    }
    /**
     * Terminate client connection
     * @param code - termination code
     * @param data - termination reason
     */
    terminate(code, data) {
        this.status = ClientStatus.disconnecting;
        this._terminate(code, data);
    }
}
exports.Client = Client;
/**
 * Abstarct class for Transport
 */
class Transport {
    constructor() {
        this.handlers = {
            connection: noop,
            disconnect: noop,
            message: noop,
            error: noop,
        };
    }
    /**
     * Register handler for client connection event
     * @param cb - connection handler
     */
    onConnection(cb) {
        this.handlers.connection = cb;
    }
    /**
     * Register handler for client message event
     * @param cb - message handler
     */
    onMessage(cb) {
        this.handlers.message = cb;
    }
    /**
     * Register handler for client disconnection event
     * @param cb - disconnection handler
     */
    onDisconnect(cb) {
        this.handlers.disconnect = cb;
    }
    /**
     * Register handler for error event
     * @param cb - error handler
     */
    onError(cb) {
        this.handlers.error = cb;
    }
}
exports.Transport = Transport;
