// wsapix/dist/transport/uws.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.uWebsocketTransport = exports.uWebSocketClient = void 0;
const transport_1 = require("./transport");
class uWebSocketClient extends transport_1.Client {
    constructor(ws) {
        super();
        this.ws = ws;
        this.path = ws.url;
        this.headers = ws.headers;
        this.query = ws.query;
    }
    _send(data, cb) {
        this.ws.send(data, false, false);
        return Promise.resolve(cb && cb());
    }
    _terminate(code, data) {
        this.ws.end(code, data);
    }
}
exports.uWebSocketClient = uWebSocketClient;
class uWebsocketTransport extends transport_1.Transport {
    constructor(options) {
        super();
        this.clients = new WeakMap();
        const { server } = options, wsOptions = __rest(options, ["server"]);
        this.app = server;
        this.app.ws('/*', Object.assign(Object.assign({}, wsOptions), { upgrade: (res, req, context) => {
                // get all headers
                const headers = {};
                req.forEach((key, value) => headers[key] = value);
                const upgradeParams = {
                    url: req.getUrl(),
                    query: req.getQuery(),
                    headers,
                    connection: {
                        remoteAddress: Buffer.from(res.getRemoteAddressAsText()).toString()
                    }
                };
                /* This immediately calls open handler, you must not use res after this call */
                /* Spell these correctly */
                res.upgrade(upgradeParams, req.getHeader('sec-websocket-key'), req.getHeader('sec-websocket-protocol'), req.getHeader('sec-websocket-extensions'), context);
            }, open: (ws) => __awaiter(this, void 0, void 0, function* () {
                const client = new uWebSocketClient(ws);
                this.clients.set(ws, client);
                client.status = transport_1.ClientStatus.connected;
                this.handlers.connection(client);
            }), close: (ws, code, message) => {
                const client = this.clients.get(ws);
                client.status = transport_1.ClientStatus.disconnecting;
                this.handlers.disconnect(client, code, Buffer.from(message.slice(0)).toString());
                client.status = transport_1.ClientStatus.disconnected;
            }, message: (ws, message, isBinary) => {
                const client = this.clients.get(ws);
                this.handlers.message(client, Buffer.from(message.slice(0)).toString());
            } }));
    }
    close(cb) {
        return Promise.resolve(cb && cb());
    }
}
exports.uWebsocketTransport = uWebsocketTransport;
