// wsapix/dist/transport/websocket.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebsocketTransport = exports.WebsocketClient = void 0;
const ws_1 = __importDefault(require("ws"));
const transport_1 = require("./transport");
const promiseCallback = (resolve, reject, cb) => {
    return (error) => {
        cb && cb(error);
        if (error) {
            return reject(error);
        }
        return resolve();
    };
};
class WebsocketClient extends transport_1.Client {
    constructor(ws, req) {
        super();
        this.ws = ws;
        const [path, query] = (req.url || "").split("?");
        this.path = path;
        this.query = query;
        this.headers = req.headers;
    }
    _send(data, cb) {
        return new Promise((resolve, reject) => {
            this.ws.send(data, promiseCallback(resolve, reject, cb));
        });
    }
    _terminate(code, reason) {
        this.ws.close(code, reason);
    }
}
exports.WebsocketClient = WebsocketClient;
class WebsocketTransport extends transport_1.Transport {
    constructor(options) {
        super();
        this.wss = new ws_1.default.Server(options);
        this.wss.on("connection", (ws, req) => {
            const client = new WebsocketClient(ws, req);
            ws.on("close", (code, data) => {
                client.status = transport_1.ClientStatus.disconnecting;
                this.handlers.disconnect(client, code, data);
                client.status = transport_1.ClientStatus.disconnected;
            });
            ws.on("message", (data) => this.handlers.message(client, data));
            client.status = transport_1.ClientStatus.connected;
            this.handlers.connection(client);
        });
        this.wss.on("error", this.handlers.error.bind(this));
    }
    close(cb) {
        return new Promise((resolve, reject) => {
            this.wss.close(promiseCallback(resolve, reject, cb));
        });
    }
}
exports.WebsocketTransport = WebsocketTransport;
