// wsapix/dist/transport/mock.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockClient = exports.MockTransport = exports.ClientSocketState = void 0;
const _1 = require(".");
const noop = () => { };
var ClientSocketState;
(function (ClientSocketState) {
    ClientSocketState[ClientSocketState["CONNECTING"] = 0] = "CONNECTING";
    ClientSocketState[ClientSocketState["OPEN"] = 1] = "OPEN";
    ClientSocketState[ClientSocketState["CLOSING"] = 2] = "CLOSING";
    ClientSocketState[ClientSocketState["CLOSED"] = 3] = "CLOSED";
})(ClientSocketState = exports.ClientSocketState || (exports.ClientSocketState = {}));
class MockTransport extends _1.Transport {
    close(cb) {
        return Promise.resolve(cb && cb());
    }
    inject(url = "/", params = {}) {
        const { headers, connectionDelay } = params, handlers = __rest(params, ["headers", "connectionDelay"]);
        const socket = {
            readyState: ClientSocketState.OPEN,
            onopen: noop,
            onerror: noop,
            onclose: noop,
            onmessage: noop,
            send: (data) => {
                this.handlers.message(client, data);
            },
            close: (code = 0, reason = "") => {
                client.status = _1.ClientStatus.disconnecting;
                setTimeout(() => {
                    this.handlers.disconnect(client, code, reason);
                    client.status = _1.ClientStatus.disconnected;
                }, connectionDelay);
            }
        };
        const client = new MockClient(socket, url, headers, connectionDelay);
        setTimeout(() => socket.onopen({ type: "open" }), connectionDelay);
        client.status = _1.ClientStatus.connected;
        this.handlers.connection(client);
        return socket;
    }
}
exports.MockTransport = MockTransport;
class MockClient extends _1.Client {
    constructor(socket, url, headers = {}, connectionDelay = 5) {
        super();
        this.socket = socket;
        this.headers = headers;
        this.connectionDelay = connectionDelay;
        const parsedUrl = new URL(url, "ws://localhost/");
        this.path = parsedUrl.pathname;
        this.query = parsedUrl.search.slice(1);
    }
    _send(data, cb) {
        return new Promise((resolve) => {
            setTimeout(() => {
                this.socket.onmessage({ type: "message", data });
                return resolve(cb && cb());
            }, this.connectionDelay);
        });
    }
    _terminate(code = 1006, reason) {
        this.socket.readyState = ClientSocketState.CLOSING;
        setTimeout(() => {
            this.socket.readyState = ClientSocketState.CLOSED;
            this.socket.onclose({ type: "close", code, reason });
        }, this.connectionDelay);
    }
}
exports.MockClient = MockClient;
