// wsapix/dist/asyncapi/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Oauth2FlowType = exports.BearerHttpSecurityScheme = exports.ApiKeyIn = exports.UserPasswordType = exports.ProtocolEnum = void 0;
var ProtocolEnum;
(function (ProtocolEnum) {
    ProtocolEnum[ProtocolEnum["http"] = 0] = "http";
    ProtocolEnum[ProtocolEnum["ws"] = 1] = "ws";
    ProtocolEnum[ProtocolEnum["amqp"] = 2] = "amqp";
    ProtocolEnum[ProtocolEnum["amqp1"] = 3] = "amqp1";
    ProtocolEnum[ProtocolEnum["mqtt"] = 4] = "mqtt";
    ProtocolEnum[ProtocolEnum["mqtt5"] = 5] = "mqtt5";
    ProtocolEnum[ProtocolEnum["kafka"] = 6] = "kafka";
    ProtocolEnum[ProtocolEnum["nats"] = 7] = "nats";
    ProtocolEnum[ProtocolEnum["jms"] = 8] = "jms";
    ProtocolEnum[ProtocolEnum["sns"] = 9] = "sns";
    ProtocolEnum[ProtocolEnum["sqs"] = 10] = "sqs";
    ProtocolEnum[ProtocolEnum["stomp"] = 11] = "stomp";
    ProtocolEnum[ProtocolEnum["redis"] = 12] = "redis";
    ProtocolEnum[ProtocolEnum["mercure"] = 13] = "mercure";
})(ProtocolEnum = exports.ProtocolEnum || (exports.ProtocolEnum = {}));
var UserPasswordType;
(function (UserPasswordType) {
    UserPasswordType["USER_PASSWORD"] = "userPassword";
    UserPasswordType["API_KEY"] = "apiKey";
    UserPasswordType["X509"] = "X509";
    UserPasswordType["SYMMETRIC_ENCRYPTION"] = "symmetricEncryption";
    UserPasswordType["ASYMMETRIC_ENCRYPTION"] = "asymmetricEncryption";
    UserPasswordType["HTTP"] = "http";
    UserPasswordType["HTTP_API_KEY"] = "httpApiKey";
    UserPasswordType["OAUTH2"] = "oauth2";
    UserPasswordType["OPEN_ID_CONNECT"] = "openIdConnect";
})(UserPasswordType = exports.UserPasswordType || (exports.UserPasswordType = {}));
var ApiKeyIn;
(function (ApiKeyIn) {
    ApiKeyIn["USER"] = "user";
    ApiKeyIn["PASSWORD"] = "password";
    ApiKeyIn["HEADER"] = "header";
    ApiKeyIn["QUERY"] = "query";
    ApiKeyIn["COOKIE"] = "cookie";
})(ApiKeyIn = exports.ApiKeyIn || (exports.ApiKeyIn = {}));
var BearerHttpSecurityScheme;
(function (BearerHttpSecurityScheme) {
    BearerHttpSecurityScheme["BEARER"] = "bearer";
})(BearerHttpSecurityScheme = exports.BearerHttpSecurityScheme || (exports.BearerHttpSecurityScheme = {}));
var Oauth2FlowType;
(function (Oauth2FlowType) {
    Oauth2FlowType[Oauth2FlowType["implicit"] = 0] = "implicit";
    Oauth2FlowType[Oauth2FlowType["password"] = 1] = "password";
    Oauth2FlowType[Oauth2FlowType["clientCredentials"] = 2] = "clientCredentials";
    Oauth2FlowType[Oauth2FlowType["authorizationCode"] = 3] = "authorizationCode";
})(Oauth2FlowType = exports.Oauth2FlowType || (exports.Oauth2FlowType = {}));
