// wsapix/dist/asyncapi/builder.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncApiBuilder = void 0;
class AsyncApiBuilder {
    constructor(params) {
        this.params = params;
        this.channels = new Map();
        this.messages = new Map();
        this.schemas = new Map();
        this.addSchemaRef = (schema) => {
            const { $id } = schema, data = __rest(schema, ["$id"]);
            if (data.type === "object") {
                Object.keys(schema.properties).forEach((key) => {
                    data.properties = Object.assign(Object.assign({}, data.properties), { [key]: this.addSchemaRef(schema.properties[key]) });
                });
            }
            else if (schema.type === "array") {
                data.items = this.addSchemaRef(schema.items);
            }
            if ($id) {
                this.schemas.set($id, data);
                return { $ref: `#/components/schemas/${$id}` };
            }
            return data;
        };
    }
    generate() {
        const map2obj = (map) => {
            const result = {};
            map.forEach((value, key) => result[key] = value);
            return result;
        };
        return JSON.stringify(Object.assign(Object.assign({ asyncapi: "2.0.0" }, this.params), { channels: map2obj(this.channels), components: {
                messages: map2obj(this.messages),
                schemas: map2obj(this.schemas)
            } }));
    }
    addChannel(name, pubMessages, subMessages, params) {
        this.channels.set(name, Object.assign(Object.assign({}, pubMessages.length ? { publish: {
                description: "Send messages to the server",
                message: Object.assign({}, pubMessages.length > 1 ? {
                    oneOf: pubMessages.map((msg) => this.addMessageRef(msg))
                } : this.addMessageRef(pubMessages[0]))
            } } : {}), subMessages.length ? { subscribe: {
                description: "Messages that you receive from the server",
                message: Object.assign({}, subMessages.length > 1 ? {
                    oneOf: subMessages.map((msg) => this.addMessageRef(msg))
                } : this.addMessageRef(subMessages[0]))
            } } : {}));
    }
    addMessageRef(msg) {
        const { $id } = msg, data = __rest(msg, ["$id"]);
        const payload = this.addSchemaRef(msg.payload);
        const message = Object.assign(Object.assign({}, data), { payload });
        if ($id) {
            this.messages.set($id, message);
        }
        return $id ? { $ref: `#/components/messages/${$id}` } : message;
    }
}
exports.AsyncApiBuilder = AsyncApiBuilder;
